### PR TITLE
Separate qc steps

### PR DESCRIPTION
**Pull request by ogibson-biomage**
# Background
#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [ ] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [ ] Deployed to staging
- [ ] Passed integration tests 

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: pipeline-runner/init.R [hunk 1]
```diff
@@ -229,25 +217,48 @@ call_data_processing <- function(task_name, input, pipeline_config) {
...
     # call function to run and update global variable
     c(
-        data, ...rest_of_results
-    ) %<-% run_processing_step(scdata, config, task_name, sample_id, debug_config)
+        data,new_ids,...rest_of_results
+    ) %<-% run_processing_step(scdata, config, tasks,task_name, cells_id,sample_id, debug_config)
+
+    print("Comparison between cell ids")
+    print(paste0("Old ids length ",length(cells_id[[sample_id]])))
+    print(paste0("New ids length ",length(new_ids[[sample_id]])))
```
Comment (ogibson-biomage) [inline pipeline-runner/init.R:245]:
```suggestion
    message("Comparison between cell ids")
    message(paste0("Old ids length ",length(cells_id[[sample_id]])))
    message(paste0("New ids length ",length(new_ids[[sample_id]])))
```

Comment (ogibson-biomage) [inline pipeline-runner/init.R:245]:
```suggestion
    message("Comparison between cell ids")
    message("Old ids length ",length(cells_id[[sample_id]]))
    message("New ids length ",length(new_ids[[sample_id]]))
```

File: pipeline-runner/R/qc-5-filter_doublets.R [hunk 2]
```diff
@@ -17,90 +17,89 @@
...
 
   return(threshold)
 }
+
+generate_doublets_plot_data <- function(scdata,num_cells_to_downsample){
+  plot1_data <- lapply(unname(scdata$doublet_scores), function(x) {
+    c("doubletP" = x)
+  })
+
+  plot1_data <- plot1_data[get_positions_to_keep(scdata,num_cells_to_downsample)]
```
Comment (alexvpickering) [inline pipeline-runner/R/qc-5-filter_doublets.R:102]: Again I would downsample first and make it consistent with `generate_MT_plot_data` where e.g. the downsampling happens outside and then `generate_doublets_plot_data` is only called after downsample

File: pipeline-runner/R/qc-helpers.R [hunk 3]
```diff
@@ -1,3 +1,42 @@
...
+#' @param scdata
+#'
+#' @return
+#' @export
+#'
+#' @examples
+generate_first_step_ids <- function(scdata){
+  cells_id <- list()
+  for(sample_id in unique(scdata$samples)){
+    cells_id[[sample_id]] <- scdata@meta.data[scdata$samples == sample_id,"cells_id"]
```
Comment (alexvpickering) [inline pipeline-runner/R/qc-helpers.R:12]:
```suggestion
    cells_id[[sample_id]] <- scdata$cells_id[scdata$samples == sample_id]
```

**Review by alexvpickering — COMMENTED** [review]
I haven't had a chance to look at the updated tests yet but have added some suggestions. Let me know if you want to discuss anything as I am happy to change my mind on anything.

File: pipeline-runner/R/handle_data.R [hunk 4]
```diff
@@ -1,16 +1,52 @@
...
+    Bucket = bucket,
     Key = paste(experiment_id, "r.rds", sep = "/")
   )
   obj <- readRDS(rawConnection(body))
   return(obj)
 }
 
+load_cells_id_from_s3 <- function(pipeline_config,task_name,experiment_id) {
+  s3 <- paws::s3(config = pipeline_config$aws_config)
+  object_list <- s3$list_objects("biomage-filtered-cells-development",Prefix=paste0(experiment_id,"/",task_name,"/"))
```
[the remark this answers is not in the document]
Comment (ogibson-biomage) [reply pipeline-runner/R/handle_data.R:32]: Added if (!sample_id %in% samples) stop("The samples in s3 don't match the samples in the object. Pipeline needs to rerun.")

We still list the objects, but we double check and stop if there's an error :)

File: pipeline-runner/R/handle_data.R [hunk 5]
```diff
@@ -1,16 +1,56 @@
...
+
+reload_scdata_from_s3 <- function(pipeline_config, experiment_id, task_name, tasks) {
+  # If the task is after data integration, we need to get scdata from processed_matrix
+  task_names <- names(tasks)
+  integration_index <- which(task_names == "dataIntegration")
+  if (match(task_name, task_names) > integration_index) {
+    bucket <- pipeline_config$processed_bucket
+  } else {
+    bucket <- pipeline_config$source_bucket
+  }
```
Comment (alexvpickering) [inline pipeline-runner/R/handle_data.R:17]: move this check outside function as we discussed

File: pipeline-runner/R/handle_data.R [hunk 6]
```diff
@@ -1,16 +1,56 @@
...
+    Bucket = bucket,
     Key = paste(experiment_id, "r.rds", sep = "/")
   )
   obj <- readRDS(rawConnection(body))
   return(obj)
 }
 
+load_cells_id_from_s3 <- function(pipeline_config, task_name, experiment_id, samples) {
+  s3 <- paws::s3(config = pipeline_config$aws_config)
+  object_list <- s3$list_objects("biomage-filtered-cells-development", Prefix = paste0(experiment_id, "/", task_name, "/"))
```
Comment (alexvpickering) [inline pipeline-runner/R/handle_data.R:32]: add a check either here or outside functions that we retrieved cell ids for all samples in scdata

File: pipeline-runner/init.R [hunk 7]
```diff
@@ -135,20 +124,10 @@ run_processing_step <- function(scdata, config, task_name, sample_id, debug_conf
     # run task and time it
     tstart <- Sys.time()
 
-    out <- task(scdata, config, sample_id, task_name)
+    out <- task(scdata, config, sample_id, cells_id,task_name)
```
Comment (alexvpickering) [inline pipeline-runner/init.R:127]:
```suggestion
    out <- task(scdata, config, sample_id, cells_id, task_name)
```

File: pipeline-runner/init.R [hunk 8]
```diff
@@ -229,25 +217,49 @@ call_data_processing <- function(task_name, input, pipeline_config) {
 
         # assign it to the global environment so we can
         # persist it across runs of the wrapper
-        assign("scdata", reload_scdata_from_s3(pipeline_config, experiment_id), pos = ".GlobalEnv")
+        assign("scdata", reload_scdata_from_s3(pipeline_config, experiment_id,task_name,tasks), pos = ".GlobalEnv")
```
Comment (alexvpickering) [inline pipeline-runner/init.R:220]:
```suggestion
        assign("scdata", reload_scdata_from_s3(pipeline_config, experiment_id, task_name, tasks), pos = ".GlobalEnv")
```

lint this file please

File: pipeline-runner/R/qc-helpers.R [hunk 9]
```diff
@@ -1,3 +1,42 @@
...
+  }
+  return(cells_id)
+}
+
+#
+# Returns the positions to keep based on scdata and number of cells to keep
+#
+get_positions_to_keep <- function(scdata, num_cells_to_downsample) {
+  # Downsample plotData
+  num_cells_to_downsample <- downsample_plotdata(ncol(scdata), num_cells_to_downsample)
```
Comment (alexvpickering) [inline pipeline-runner/R/qc-helpers.R:22]: do we use `downsample_plotdata` anywhere else still? If not, might make sense to just pull it into this function as its pretty small and poorly named

File: pipeline-runner/R/qc-helpers.R [hunk 10]
```diff
@@ -1,3 +1,42 @@
...
+  cells_position_to_keep <- sample(1:ncol(scdata), num_cells_to_downsample, replace = FALSE)
+  cells_position_to_keep <- sort(cells_position_to_keep)
+
+  return(cells_position_to_keep)
+}
+
+#
+# subset_ids subsets a seurat object with the cell ids
+#
+subset_ids <- function(scdata, cells_id) {
```
Comment (alexvpickering) [inline pipeline-runner/R/qc-helpers.R:33]: wondering if we should differentiate the fact that `cells_id` is usually a list of vectors of `cells_id` for each sample but here it is just a vector of `cells_id`. Might want to rename either here or elsewhere.

File: pipeline-runner/tests/testthat/qc_mock.R [hunk 11]
```diff
@@ -0,0 +1,9 @@
+#' Title
+#'
+#' @return
+#' @export
+#'
+#' @examples
```
Comment (alexvpickering) [inline pipeline-runner/tests/testthat/qc_mock.R:6]:
```suggestion
```

no need to have un-filled doc strings

**Pull request merged**